Updated derivatives for complex mm, mv, ger, bmm, triangular_solve

IvanYashchuk opened the pull request:
This PR updates derivatives for a few functions so that `gradgradcheck` for `torch.cholesky` is passed ([ref](https://github.com/pytorch/pytorch/pull/45267#discussion_r494439967)).

Some tests (that call to `bmm_cuda`) fail with with `RuntimeError: _th_bmm_out not supported on CUDAType for ComplexDouble`
until PR #42553 is merged.

Ref. #33152

IvanYashchuk committed:
Updated backward for complex triangular solve

IvanYashchuk committed:
Updated backward for complex matmul

IvanYashchuk committed:
Updated backward for ger and mv

IvanYashchuk committed:
Updated backward for complex bmm

IvanYashchuk committed:
Now gradgradcheck is working for complex triangular_solve

IvanYashchuk committed:
Allow diagonal for complex autograd.

Enable gradgradcheck for complex cholesky

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into complex-grad

anjali411 commented:
@IvanYashchuk this PR looks good overall. Sorry I didn't get a chance to finish working on #42553 because of complex autograd issues. I would try to work on it this week. In case, it's blocking a lot of other work, would you like to take over #42553?

anjali411 commented on test/test_autograd.py:
```diff
@@ -4833,7 +4832,8 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
                 'eq_', 'ne_', 'add', '__radd__', 'sum', 'conj', 'sin', 'cos', 'mul', 'sinh',
-                'cosh', '__rmul__', 'sgn', 'abs', 'dot', 'vdot', 'tensor_split'] + separate_complex_tests
+                'cosh', '__rmul__', 'sgn', 'abs', 'dot', 'vdot', 'tensor_split', 'matmul', 'mv',
+                'ger', 'bmm', 'diagonal', ] + separate_complex_tests
```
I think we can also add tests for related functions: `addmv`, `addbmm`

anjali411 replied:
@IvanYashchuk can you remove `bmm` from this list?

IvanYashchuk committed:
Removed bmm from complex_list test

Correctness of the derivatiev rule for bmm is being implicitly tested in
triangular_solve for now.

IvanYashchuk committed:
Skip failing complex cuda test cases for bmm and matmul

We should remember to re-enable them once batched matmul for complex is
supported on cuda

anjali411 commented on test/test_autograd.py:
```diff
@@ -4833,7 +4832,11 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
+                'cosh', '__rmul__', 'sgn', 'abs', 'dot', 'vdot', 'tensor_split',
+                'matmul', 'bmm', 'mv', 'ger', 'diagonal', ] + separate_complex_tests
+
+# this list corresponds to cases that are not currently implemented
```
nit - currently implemented for complex

anjali411 commented on tools/autograd/derivatives.yaml:
```diff
@@ -749,8 +749,8 @@
-  self: grad.ger(vec)
-  vec: self.t().mv(grad)
+  self: grad.ger(vec.conj())
+  vec: self.t().conj().mv(grad)
```
nit - `self.conj().t().mv(grad)` for better readability

IvanYashchuk commented on test/test_autograd.py:
```diff
@@ -5019,6 +5022,11 @@ def fn(*inputs):
                 do_test = skip(do_test)
 
+            # TODO: remove this once tests from skip_cuda_list work
+            do_test = skipCUDAIf(
```
This doesn't skip tests for floating-point dtype, because it specifically skips only those tests that match `test_name` with any in `skip_cuda_list = ['bmm_complex', 'matmul_4d_4d_complex']` which have `_complex` in the name.

anjali411 replied:
ohh right my bad

IvanYashchuk committed:
Reorder t() conj()

anjali411 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -530,9 +530,9 @@ Tensor mm_mat1_backward(const Tensor & grad, const Tensor & mat2, const Tensor &
+    return maybe_multiply(mat2.conj().mm(grad.t()).t(), alpha);
   } else {
-    return maybe_multiply(grad.mm(mat2.t()), alpha);
+    return maybe_multiply(grad.mm(mat2.t().conj()), alpha);
```
nit - `grad.mm(mat2.conj().t())`

anjali411 reviewed (approved):
LGTM thanks @IvanYashchuk
